<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [20]:
import pandas as pd
import psycopg2

In [56]:
# connect

In [57]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [25]:
# текст запроса
query_3_1 = f'''
select count(id)
from vacancies
'''
df = pd.read_sql_query(query_3_1, connection)
df

/tmp/ipykernel_144082/3622974877.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

# Результат запроса
В базе содержится 49197 вакансий.

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [10]:
# текст запроса
query_3_2 = f'''
select count(id)
from employers
'''
df = pd.read_sql_query(query_3_2, connection)
df

/tmp/ipykernel_144082/1560299513.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

# Результат запроса
В базе данных 23501 работодателей.

3. Посчитате с помощью запроса количество регионов (таблица с).

In [11]:
# текст запроса
query_3_3 = f'''
select count(id)
from areas
'''
df = pd.read_sql_query(query_3_3, connection)
df

/tmp/ipykernel_144082/3625548944.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

# Результат запроса
 В базе данных 1362 региона.

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [12]:
# текст запроса
query_3_4 = f'''
select count(id)
from industries
'''
df = pd.read_sql_query(query_3_4, connection)
df

/tmp/ipykernel_144082/4178910013.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

# Результат запроса
В базе данных 294 сферы деятельности.

***

# выводы по предварительному анализу данных
В базе данных содержится 49197 вакансий от 23501 работодателя для 294 сфер деятельности в 1362 регионах.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
query_4_1 = f'''
select  a.name, count(v.id) cnt
from vacancies v join areas a on a.id=v.area_id
group by a.name
order by cnt desc
'''
df = pd.read_sql_query(query_4_1, connection)
df

/tmp/ipykernel_144082/3590210995.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

# Результат запроса
В пятерку лидеров по количеству вакансий входят следующие регионы: Москва, Санкт-Петербург, Минск, Новосибирск и Алматы.

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''
select count(id)
from vacancies
where salary_from is not null or salary_to is not null
'''
df = pd.read_sql_query(query_4_2, connection)
df


/tmp/ipykernel_144082/3895430518.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

# Результат запроса
В базе данных 24703 вакансии с заполненной информацией о зарплате, что составляет 50,21% от общего числа вакансий.

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [15]:
# текст запроса
query_4_3 = f'''
select round(avg(salary_from)) as start_salary,
round(avg(salary_to)) as end_salary
from vacancies
'''
df = pd.read_sql_query(query_4_3, connection)
df

/tmp/ipykernel_144082/451719518.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


start_salary  end_salary
0       71065.0    110537.0

# Результат запроса
Среднее значение нижней границы зарплатной вилки составляет 71065 руб., в то время как среднее значение верхней границы 110537 руб.

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [16]:
# текст запроса
query_4_4 = f'''
select schedule, employment, count(id)
from vacancies
group by schedule, employment
order by count(id) desc
'''
df = pd.read_sql_query(query_4_4, connection)
df

/tmp/ipykernel_144082/1646781780.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

# Результат запроса
Лидирует по типу трудоустройства - "полная занятость": около 94% от всех вакансий, оставшиеся 6% делят "частичная занятость" и стажировка, в то время как вакансий на проектную работу незначительное количество (менее 0.04 %).
По типу рабочего графика "полный рабочий день" требуется в 36424 вакансиях,"удаленная работа" в 9311, а "гибкий график" -2039.
Большинство работодателей все ещё предпочитают нанимать сотрудников на полную занятость с графиком работы полный рабочий день - 35367. Несмотря на это мы видим проникновение мировых трендов на рынок вакансий: на втором месте удаленная работа с полной занятостью -7802. Вероятно это обусловлено пандемией, желанием работодателей сократить операционные расходы, экономией времени и удобством для соискателей.

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [22]:
# текст запроса
query_4_5 = f'''
select experience, count(id)
from vacancies
where experience is not null
group by experience
order by count(id) desc
'''
df = pd.read_sql_query(query_4_5, connection)
df

/tmp/ipykernel_144082/3750186481.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0  От 1 года до 3 лет  26152
1       От 3 до 6 лет  14511
2           Нет опыта   7197
3         Более 6 лет   1337

# Результат запроса
Наибольшее количество вакансий требует начальных знаний и опыта от 1-3 лет. Работодатели более  заинтересованы в привлечении молодых специалистов, чем более опытных кандидатов. Вероятно, специалисты с опытом более 6 лет в индустрии достаточно известны, либо чаще используют нетворкинг и другие каналы поиска работы.

***

In [ ]:
# выводы по детальному анализу вакансий

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [20]:
# текст запроса
query_5_1 = f'''
select e.name,count(v.id) as number
from vacancies v
inner join employers e on e.id=v.employer_id
group by  e.name
order by count(v.id) desc
LIMIT 5
'''
df = pd.read_sql_query(query_5_1, connection)
df

/tmp/ipykernel_144082/2956601185.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  number
0         Яндекс    1933
1     Ростелеком     491
2       Тинькофф     444
3           СБЕР     428
4  Газпром нефть     331

# Pезультат запроса
У компании Яндекс наибольшее количество открытых вакансий - 1933, в то время как Газпром нефть замыкает пятерку лидеров с 331 открытой вакансией.

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [29]:
# текст запроса
query_5_2 = f'''
select a.name, count(e.id) as employer,count(v.id) as vacancies
from areas a
left join employers e on a.id=e.area
left join vacancies v on v.area_id=a.id

group by a.name
order by vacancies, employer desc
'''
df = pd.read_sql_query(query_5_2, connection)
df

/tmp/ipykernel_144082/2659661473.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  employer  vacancies
0                 Россия       410          0
1              Казахстан       207          0
2     Московская область        75          0
3     Краснодарский край        19          0
4     Ростовская область        18          0
...                  ...       ...        ...
1357         Новосибирск   1149438    1149438
1358              Алматы   1364132    1364132
1359               Минск   2354880    2354880
1360     Санкт-Петербург   6320667    6320667
1361              Москва  31272712   31272712

[1362 rows x 3 columns]

# Результат запроса
Наибольшее количество работодателей при отсутствии открытых вакансий зарегистрировано в России - 410, также на рынке предствалены работодатели из других стран, например, Австралии, Сингапура и Канады.

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [33]:
# текст запроса

query_5_3= f'''
select v.employer_id, e.name, count(distinct v.area_id)
from vacancies v
join employers e on v.employer_id=e.id
group by employer_id,e.name
order by count(distinct area_id) desc
'''
df = pd.read_sql_query(query_5_3, connection)
df

/tmp/ipykernel_144082/4101608865.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employer_id                                    name  count
0             1740                                  Яндекс    181
1             2748                              Ростелеком    152
2          5724811                              Спецремонт    116
3          5130287                  Поляков Денис Иванович     88
4          3682876                               ООО ЕФИН      71
...            ...                                     ...    ...
14901       810278                            НПП Авиатрон      1
14902       810313              Центр дистанционных торгов      1
14903       810551  Городские Телекоммуникационные Системы      1
14904       810688                       Введенский, Отель      1
14905       811902              Группа Компаний Автоимпорт      1

[14906 rows x 3 columns]

# Результат запроса
У компании Яндекс открыты вакансии в наибольшем количестве регионов - 181, далее следует Ростелеком -152 региона.

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [7]:
# текст запроса
query_5_4= f'''
select count(*) from employers WHERE id not IN (select employer_id
from employers_industries)
'''
df = pd.read_sql_query(query_5_4, connection)
df

/tmp/ipykernel_312094/294244015.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

# Результат запроса
У 8419 работодателей не указана сфера деятельности.

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [11]:
# текст запроса
query_5_5= f'''
select e.name, count(ei.industry_id)
from employers_industries ei
join employers e on e.id=ei.employer_id
group by e.name
having count(ei.industry_id)=4
order by e.name asc
'''
df = pd.read_sql_query(query_5_5, connection)
df

/tmp/ipykernel_312094/2389486858.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0        101 Интернет      4
1            21vek.by      4
2                2ГИС      4
3                  2К      4
4         4 пикселя +      4
...               ...    ...
1131            ЮРИОН      4
1132  ЮТИП Технологии      4
1133           ЯКласс      4
1134            ЯрНео      4
1135    Яшин&Партнёры      4

[1136 rows x 2 columns]

# Результат запроса
Многие компании выкладывают вакансии в 4 и более индустриях, одна из таких компаний 2ГИС(третья в алфавитном порядке)

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [18]:
# текст запроса
query_5_6= f'''
select count(ei.employer_id)
from employers_industries ei
join industries i on i.id=ei.industry_id
group by i.name
having i.name ='Разработка программного обеспечения'
'''
df = pd.read_sql_query(query_5_6, connection)
df

/tmp/ipykernel_312094/3647077534.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

# результат запроса
3553 компании ищут IT специалистов, т.е выкладывают вакансии в категории "Разработка программного обеспечения".

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [5]:
# код для получения списка городов-милионников
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup
# Определяем значение URL страницы для запроса
url ='https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
city_million=[x.text for x in page.find_all('div', class_="ts-ПозКарта-метка-ссылка")]
print(city_million)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Красноярск', 'Екатеринбург', 'Нижний Новгород', 'Самара', 'Казань', 'Пермь', 'Омск', 'Челябинск', 'Ростов-на-Дону', 'Волгоград', 'Уфа', 'Воронеж', 'Краснодар']


In [12]:
# текст запроса
query_5_7= f'''

with tmp_select as (
 select a.name, count(v.id) as "vacancies_count"
  from vacancies v
  join employers e on v.employer_id=e.id
  join areas a on a.id=v.area_id
  where e.name='Яндекс' and (a.name in('{"','".join(city_million)}'))
  group by v.area_id,a.name
)

(select * from tmp_select)
union all
(select 'Total' as name, sum("vacancies_count") from tmp_select)

'''
df = pd.read_sql_query(query_5_7, connection)
df

/tmp/ipykernel_339594/3352129555.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  vacancies_count
0            Москва             54.0
1   Санкт-Петербург             42.0
2      Екатеринбург             39.0
3       Новосибирск             35.0
4         Волгоград             24.0
5           Воронеж             32.0
6         Краснодар             30.0
7        Красноярск             23.0
8   Нижний Новгород             36.0
9              Омск             21.0
10            Пермь             25.0
11   Ростов-на-Дону             25.0
12           Самара             26.0
13           Казань             25.0
14              Уфа             26.0
15        Челябинск             22.0
16            Total            485.0

# Результат запроса
У компании "Яндекс" 285 открытых вакансий в 16 городах-миллионниках.

***

# Выводы по анализу работодателей
На сайие предствавлено 23175 работодателей,которые ищут кандидатов на 49197 вакансий в различныйх регионах и странах.
Подавляющее большинство вакансий открыто в России, но также есть малый процент вакансий для Австралии, Сингапура и Канады.
У более чем тысячи работодателей указано 4 и более сферы деятельности, но есть  и 8419 работодателей оставивших это поле пустым.
Наибольшее количество кандидатов требуется крупным компаниям таким как Яндекс, Ростелеком, Тинькофф и СБЕР.
Например, компания Яндекс ищет сотрудников в 181 регионе России, в то время как Ростелеком только в 104.
У Яндекса 1933 открытые вакансии, но только 285 из них привязаны к городам-миллионникам, что свидетельствует о большом смещении фокуса на удаленную работу.
Также стоит отметить, что наблюдается острая нехватка кадров в IT сфере - 3553 открытых вакансии.



# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [4]:
# текст запроса
query_6_1= f'''
select count(* )
from vacancies v
where ((LOWER(v.name) like '%data%')or(LOWER(v.name) like '%данн%'))

'''
df = pd.read_sql_query(query_6_1, connection)
df

/tmp/ipykernel_19811/270794917.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

# Pезультат запроса
На сайте открыта 1771 вакансия, связанная с данными.

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [17]:
# текст запроса
query_6_2= f'''
with DS as (
select *
from vacancies v
where (lower(name) like any (array['%data scientist%','%data science%','%исследователь данных%','% ml%','%/ml%','%/ ml%','%machine learning%','%машинн%обучен%']))
)
(select count(*)
from DS
where (lower(name) like '%junior%')or (experience='Нет опыта')or(employment='Стажировка')

)
'''
df = pd.read_sql_query(query_6_2, connection)
df

/tmp/ipykernel_19811/1665392559.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

# Результат запроса
На текущий момент на рынке есть 51 вакансия для начинающего дата-саентиста.

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [18]:
# текст запроса
query_6_3= f'''
select count(mlv.name)
from (
  select * from vacancies v
  where (lower(name) like any (array['%data scientist%','%data science%','%исследователь данных%','% ml%','%/ml%','ml%','%machine learning%','%машинн%обучен%']))
) as mlv
where mlv.key_skills IS NOT NULL AND (mlv.key_skills like '%SQL%' or lower(mlv.key_skills) LIKE '%postgre%')
'''
df = pd.read_sql_query(query_6_3, connection)
df

/tmp/ipykernel_19811/401874473.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    198

# Результат запроса
198 вакансий для дата-саентистов требуеют хорошего знания языка запросов SQL.

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [25]:
# текст запроса
query_6_4= f'''
select count(mlv.name)
from (
  select * from vacancies v
  where (lower(name) like any (array['%data scientist%','%data science%','%исследователь данных%','%machine learning%','%машинн%обучен%']))or(name like '%ML%' and name not like'%HTML%')
) as mlv
where mlv.key_skills IS NOT NULL AND ( mlv.key_skills ilike '%Python%')
'''
df = pd.read_sql_query(query_6_4, connection)
df

/tmp/ipykernel_19811/1225807955.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

# Результат запроса
Знание языка программирования Python требутся в 351 вакансии, что в 1,8 раза чаще чем требования к знанию SQL.

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [53]:
# текст запроса
query_6_5= f'''
select  round(avg(length(key_skills)
             -length(replace(key_skills, CHR(9), ''))+1), 2)
 from vacancies v
  where (lower(name) like any (array['%data scientist%','%data science%','%исследователь данных%','%machine learning%','%машинн%обучен%']))or(name like'%ML%' and name not like'%HTML%')

and(key_skills IS NOT NULL )

'''
df = pd.read_sql_query(query_6_5, connection)
df

/tmp/ipykernel_19811/1673273273.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

# Результат запроса
Профессия DS требует огромного спектра знаний: среднее количество требуемых ключевых навыков более 6

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [58]:
# текст запроса
query_6_6= f'''
with DS as (select  *
 from vacancies v
  where (lower(name) like any (array['%data scientist%','%data science%','%исследователь данных%','%machine learning%','%машинн%обучен%']))or(name like'%ML%' and name not like'%HTML%')

and(key_skills IS NOT NULL )
)
select round(AVG(coalesce((salary_from+salary_to)/2, salary_from, salary_to)),0)
from DS
where (experience ='От 3 до 6 лет') and ((salary_from is not null) or(salary_to is not null))

'''
df = pd.read_sql_query(query_6_6, connection)
df

/tmp/ipykernel_19811/783395382.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


round
0  243115.0

# Результат запроса
По данным РосСтат на 2021 год средняя зарплата по России составляет 57 244 руб. Опытный дата-сайнтист , имеющий от 3 до 6 лет опыта, может в среднем зарабатывать в 4.24 раза больше.

***

# выводы по предметному анализу
Открыта 1771 вакансия, связанная с данными. Несмотря на то, что профессия дата-саентиста требует широкого кругозора и большого количества различных умений (более 6 различных ключевых навыков), на рынке всё же присутствует 51 вакансия для начинающих специалистов.
Python как ключевой навык требуется в 1.8 раза чаще, чем SQL, так как он является основным языком ML разработчиков.
Средняя заработная плата специалиста составляет 243115 руб., что в 4.24 раза выше средней зарплаты по России. Такая высокая ЗП обусловлена большим количеством знаний и требований.

# Общий вывод по проекту

Количество вакансий на сайте HeadHunter составляет 49197. 23501 работодатель ищут кандидатов в 1362 регионах и 294 сферах деятельности.
Большинство вакансий открыты в городах-миллионниках: Москве, Санкт-Петербурге, Минске, Новосибирске и Алматы. Преимущественно этот ресурс ориентирован на русскоязычных кандидатов из России и стран СНГ, но несмотря на это присутствуют вакансии и для других стран: Австралии, Сингапура и Канады.
Наибольшее количество кандидатов требуется крупным компаниям: Яндекс, Ростелеком, Тинькофф и СБЕР. Например, у компании Яндекс 1933 открытые вакансии, но только 285 из них привязаны к городам-миллионникам, что свидетельствует о большом смещении фокуса на удаленную работу. Также стоит отметить, что наблюдается острая нехватка кадров в IT сфере - 3553 открытых вакансии.
Одним из востребованных и перспективных направлений в IT является data science. Развитие данного направления связано с активным использованием big data, математических моделей и машинного обучения при анализе данных. Высокая заработная плата таких специалистов привлекает большое количество кандидатов. Однако, данная профессия требует большого количества компетенций (в среднем более 6 ключевых навыков), например, программирование на Python, SQL, знания математических моделей и т.д. Квалифицированных кадров на текущий момент недостаточно, что подтверждает 1771 открытая вакансия.

